runtime/Cpp/runtime/src/atn/PredictionContext.cpp

```diff
@@ -252,130 +230,123 @@ Ref<PredictionContext> PredictionContext::mergeArrays(const Ref<ArrayPredictionC
   const Ref<ArrayPredictionContext> &b, bool rootIsWildcard, PredictionContextMergeCache *mergeCache) {
 
   if (mergeCache != nullptr) {
-    auto existing = mergeCache->get(a, b);
-    if (existing) {
-      return existing;
-    }
-    existing = mergeCache->get(b, a);
-    if (existing) {
+    // This is different to the Java runtime. Merge cache lookups are symmetric.
+    if (const Ref<PredictionContext>& existing = mergeCache->get(a, b)) {
       return existing;
     }
   }
 
   // merge sorted payloads a + b => M
-  size_t i = 0; // walks a
-  size_t j = 0; // walks b
-  size_t k = 0; // walks target M array
-
-  std::vector<size_t> mergedReturnStates(a->returnStates.size() + b->returnStates.size());
-  std::vector<Ref<PredictionContext>> mergedParents(a->returnStates.size() + b->returnStates.size());
-
-  // walk and merge to yield mergedParents, mergedReturnStates
-  while (i < a->returnStates.size() && j < b->returnStates.size()) {
-    Ref<PredictionContext> a_parent = a->parents[i];
-    Ref<PredictionContext> b_parent = b->parents[j];
-    if (a->returnStates[i] == b->returnStates[j]) {
-      // same payload (stack tops are equal), must yield merged singleton
-      size_t payload = a->returnStates[i];
-      // $+$ = $
-      bool both$ = payload == EMPTY_RETURN_STATE && a_parent && b_parent;
-      bool ax_ax = (a_parent && b_parent) && *a_parent == *b_parent; // ax+ax -> ax
-      if (both$ || ax_ax) {
-        mergedParents[k] = a_parent; // choose left
-        mergedReturnStates[k] = payload;
-      }
-      else { // ax+ay -> a'[x,y]
-        Ref<PredictionContext> mergedParent = merge(a_parent, b_parent, rootIsWildcard, mergeCache);
-        mergedParents[k] = mergedParent;
-        mergedReturnStates[k] = payload;
+  std::vector<PredictionContextItem> contexts;
+  { // This scope causes iterators to be deleted before ArrayPredictionContext is constructed from the contexts vector.
+    contexts.reserve(a->contexts.size() + b->contexts.size());
+    std::vector<PredictionContextItem>::const_iterator i = a->contexts.begin(), // walks a
+                                                       iEnd = a->contexts.end();
+    std::vector<PredictionContextItem>::const_iterator j = b->contexts.begin(), // walks b
+                                                       jEnd = b->contexts.end();
+    while (i != iEnd && j != jEnd) {
+      if (i->returnState == j->returnState) {
+        // same payload (stack tops are equal), must yield merged singleton
+        size_t payload = i->returnState;
+        // $+$ = $
+        if (i->parent && j->parent && (payload == EMPTY_RETURN_STATE || *i->parent == *j->parent)) {
```

Comment:
@mike-lischke and @niccroad, This line replicates the bug in line 281 which would have stated `bool both$ = payload == EMPTY_RETURN_STATE && !a_parent && !b_parent;`  So one suggestion for fixing this bug in this version would be to have the following line 252 be modified to:
```
bool both$ = payload == EMPTY_RETURN_STATE && !(i->parent) && !(j->parent);
bool ax_ax = (i->parent && j->parent) && *i->parent == *j->parent; // ax+ax -> ax
if (both$ || ax_ax) {
```